src/GafferScene/SetAlgo.cpp

```diff
@@ -308,6 +318,11 @@ struct AstHasher
 		boost::apply_visitor( *this, expr.right.expr );
 	}
 
+	void operator()( const Nil &nil )
+	{
+		m_hash.append( false );
```

Comment:
> One of the tests in SetFilterTest runs the node with an empty string as the set name/expression. I had to explicitly hash a dummy value here, so that Gaffer doesn't throw an exception at me because SetFilter::hash() never actually hashes anything in that case. Do you have a better idea?

I don't think the SetAlgo change is the answer here. The clue is in the [documentation for ComputeNode::hash()](https://github.com/GafferHQ/gaffer/blob/master/include/Gaffer/ComputeNode.h#L68) - you need to make sure you call your base class hash before appending your own, in which case it's fine to append nothing. The base class hash will end up at the top of the hierarchy in [`ComputeNode::hash()`](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ComputeNode.cpp#L56), where the TypeId for the node and the name of the plug are used to start the hash with something unique to the node/plug combo.

> I guess we should create an empty PathMatcher for these cases after all? ;) You've been saying that all along, I think.

Oh, I thought we were doing! Yes, given that the SetFilter has been doing that all along without any complaint from users, I think we can say that it's the right thing to do. Note that this means you can entirely remove the `setNames()` check from SetAlgo, since it's the responsibility of the `SceneNode::computeSet()` to return an empty set where the name is invalid anyway.